pulsar-common/src/main/java/org/apache/pulsar/common/protocol/schema/SchemaHash.java

```diff
@@ -55,15 +82,118 @@ public static SchemaHash of(SchemaData schemaData) {
     }
 
     public static SchemaHash of(SchemaInfo schemaInfo) {
-        return of(schemaInfo == null ? new byte[0] : schemaInfo.getSchema(),
+        return of(schemaInfo == null ? null : schemaInfo.getSchema(),
                 schemaInfo == null ? null : schemaInfo.getType());
     }
 
-    public static SchemaHash of(byte[] schemaBytes, SchemaType schemaType) {
-        return new SchemaHash(hashFunction.hashBytes(schemaBytes), schemaType);
+    public static SchemaHash of() {
+        return of(null, null);
+    }
+
+    private static SchemaHash of(byte[] schemaBytes, SchemaType schemaType) {
+        SchemaHash result = null;
+        if (schemaBytes == null || schemaBytes.length == 0) {
+            result = EmptySchemaHashFactory.get(schemaType);
+        }
+
+        // This should not be a common occurrence if everything goes well.
+        if (result == null) {
+            log.warn("Could not get schemaHash from EmptySchemaHashFactory, will create by hashFunction. Might bring"
+                            + " performance regression. schemaBytes length:{}, schemaType:{}",
+                    schemaBytes == null ? "null" : schemaBytes.length, schemaType);
+            result = new SchemaHash(
+                    hashFunction.hashBytes(schemaBytes == null ? new byte[0] : schemaBytes), schemaType);
+        }
+        return result;
     }
 
     public byte[] asBytes() {
         return hash.asBytes();
     }
+
+    private static class EmptySchemaHashFactory {
+        private static final HashCode EMPTY_HASH = hashFunction.hashBytes(new byte[0]);
+        private static final SchemaHash NONE_SCHEMA_HASH = new SchemaHash(EMPTY_HASH, NONE);
+        private static final SchemaHash STRING_SCHEMA_HASH = new SchemaHash(EMPTY_HASH, STRING);
+        private static final SchemaHash JSON_SCHEMA_HASH = new SchemaHash(EMPTY_HASH, JSON);
```

Comment:
We only need to handle the primitive schema? The schema data of the struct schema is not empty, we can't use EMPTY_HASH here.

Reply:
OK. I have updated this patch. 
Only when`schemaBytes == null || schemaBytes.length == 0` will get from `EmptySchemaHashFactory`, so it doesn't matter if it is a primitive schema or not.